[RFC] Improve executable() and exepath() on windows

erw7 opened the pull request:
- [x] Fix problem that `exepath()` does not append extensions when extension is omitted(fixes #9403).
- [x] Change `executable()` to operate equivalent to vim 8.1.453(vim/vim#3512)(partially fixes #9391  that `executable()` returns 1 even if it is not executable on Windows.).
- [x] Fix an issue where `executable()` was also finding files with paths relative to the path in `$PATH`.
- [x] Write test.

Note: This pull request will not change `executable()` to use file extension and association.

erw7 committed:
win: executable(): fix relative path bug

Qualified (i.e. dot-prefixed) relative paths should only search CWD, not
$PATH.

erw7 committed:
win: executable(): also check extension

erw7 committed:
win: exepath(): append extension if omitted

fixes #9403

erw7 committed:
test/win: executable(), exepath() #9516

janlazo commented on src/nvim/testdir/test_functions.vim:
```diff
@@ -869,8 +869,8 @@ func Test_Executable()
-    call assert_equal(1, executable('shell32.dll'))
-    call assert_equal(1, executable('win.ini'))
+    call assert_equal(0, executable('shell32.dll'))
+    call assert_equal(0, executable('win.ini'))
```
What does `executable` mean on cmd.exe and sh?
On cmd.exe, running `win.ini` opens notepad on Windows so `executable` should return 1.
On sh, it does not run (because it checks the first line of the file) so it should return 0.

erw7 replied:
`win.ini` is not executable(executable command or executable program), it is a file associated with executable.

@janlazo Do you really think that you would like to search for `foo.txt` or `bar.jpg` using `executable()` from `$PATH`?

As for the UNIX like shell, it is only a current implementation due to the trade-off between implementation cost and usefulness obtained by it.

erw7 replied:
If the user associates perl with `.pl` file and adds `.pl` to `PATHEXT`, the user thinks that `.pl` file should be treated as an executable command or executable program. Therefore, `executable()` must also search for `.pl` files.

janlazo replied:
> Do you really think that you would like to search for foo.txt or bar.jpg using executable() from $PATH?

Yes. If I set `foo.txt` file permissions with Read+Execute and added its directory to `$PATH`, then I expect `executable` to find it and check if it is associated with `.exe` file because that is `executable` enough for me.

I asked what `executable` means now because this PR seems to be for Vim compatibility.

janlazo replied:
I add `.dll` to `PATHEXT`. Is `foo.dll` executable now? cmd.exe cannot run that (by default).

janlazo commented on src/nvim/os/fs.c:
```diff
@@ -254,28 +258,88 @@ bool os_can_exe(const char_u *name, char_u **abspath, bool use_path)
-/// Returns true if `name` is an executable file.
-static bool is_executable(const char *name)
+#ifdef WIN32
+/// Returns true if extension of `name` is executalbe file exteinsion.
```
typo

erw7 committed:
cleanup: PATHEXT function

wsdjeg commented:
@justinmk when this PR is merged, how could I check if my neovim has this patch?

in vim, we can use:

```vim
if has('patch-8.1.xxx')
   " use the executable()
else
  " warning or reimplement compatble api
end
```

but in neovim, how can we do this?

erw7 committed:
is_extension_executable: simplify

Simplify method of determining search position of the extension.

justinmk commented:
@erw7 would you mind rebasing this?

erw7 commented:
@justinmk I rebased this.

Although I did not create a pull request, [there](https://github.com/neovim/neovim/compare/master...erw7:improve-executable-on-windows-by-file-assoc) is an experimental  implementation of `executable()` that uses file association. Should I make a pull request for this too?

I was working on the [ACL-based libuv's `uv_fs_access`](https://github.com/libuv/libuv/compare/v1.x...erw7:improve-uv-fs-access-on-windows) implementation for #9683. And it can also be used with `executable()`. #9683 has been closed, is it worth making a pull request for it?

justinmk commented:
@erw7 thanks!

> [there](https://github.com/neovim/neovim/compare/master...erw7:improve-executable-on-windows-by-file-assoc) is an experimental implementation of `executable()` that uses file association. Should I make a pull request for this too?

I think it should be "preserved" by pasting the patch into a commit message, for historical reference. 

> Note that: If we open a file that does not contain code to execute, I think it should be:
>     elseif has('win32')
>     let s:open_prg = 'start'

Agreed, this is the role of `gx`. Also, file association lookup can be _very_ slow, can't it?

> I was working on the [ACL-based libuv's `uv_fs_access`](https://github.com/libuv/libuv/compare/v1.x...erw7:improve-uv-fs-access-on-windows) implementation for #9683. And it can also be used with `executable()`. #9683 has been closed, is it worth making a pull request for it?

I closed #9683 as duplicate of #1200, so it's still on the roadmap. Will you create a PR for upstream libuv? We can put it in our fork, while waiting for them.

justinmk committed:
is_extension_executable: simplify check for unix-style shell

mch_expand_wildcards does it this way, it's probably good enough.

justinmk commented:
- Squashed some commits
- Changed `is_extension_executable` to use simpler approach for detecting Unix-style 'shell'

justinmk commented:
@erw7 `is_executable_ext` seems redundant with `is_executable_ext`. Why not fix `is_executable_ext` instead?

justinmk committed:
fs.c: eliminate is_extension_executable

Add this functionality to is_executable_ext() instead.

justinmk commented:
@erw7 `is_executable_ext` seems redundant with `is_executable_ext`. I think we can  add this functionality to `is_executable_ext` . Please check the latest commit: b464156.

janlazo commented on test/functional/eval/executable_spec.lua:
```diff
@@ -136,21 +136,62 @@ describe('executable() (Windows)', function()
+      eq(1, call('executable', 'test_executable_'..ext..'.'..ext))
+    end
+    eq(1, call('executable', 'test_executable_zzz.zzz'))
+    clear({env={PATHEXT='.zzz;.'}})
```
Is this a repeat of the first test? Both set the shell to `sh`.

erw7 replied:
It is strange to do set shell, so it was fixed. If we put it in a loop, I don't like to do that because I don't know which pattern failed.

erw7 committed:
fs.c: fix is_executable_ext

- Fix the problem of checking the extension in a UNIX like shell.
- Fix the problem of not checking the existence of the file when the
  pathext contains an extension.

erw7 committed:
test/win: Remove unnecessary set shell

erw7 commented:
@justinmk As there was a point that I thought was a 2 point problem, I corrected it.

erw7 committed:
fs.c: Simplify calling is_executable_ext

erw7 commented:
@jusinmk After reviewing `is_executable ext`, I find that it can be changed to cc6d3f7. I think we can integrate `is_executable_ext` functionality with `is_executable`. Should we do that?

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -286,35 +282,55 @@ static bool is_executable(const char *name)
-    if (is_executable(os_buf)) {
+    size_t ext_len = (size_t)(ext_end - ext);
+    STRLCPY(buf_end, ext, ext_len + 1);
+    bool in_pathext = (strstr((char *)path_tail(p_sh), "sh") != NULL)
```
could move the "sh" check outside the loop:

    bool is_unix_shell = strstr((char *)path_tail(p_sh), "sh") != NULL;

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -351,15 +367,6 @@ static bool is_executable_in_path(const char_u *name, char_u **abspath)
-  if (!pathext) {
-    pathext = ".com;.exe;.bat;.cmd";
-  }
-  buf_len += strlen(pathext);
```
Ok to remove this because `is_executable_ext` works with `os_buf` so  `buf` doesn't need extra space.

erw7 committed:
fs.c: Move sh check of is_executable_ext to outside of loop

erw7 committed:
[ci skip] fs.c: fix comment

justinmk commented:
> I think we can integrate `is_executable_ext` functionality with `is_executable`. Should we do that?

Better leave it as-is unless there is a clear and obvious reason to combine them. I think the current separation helps a bit with readability. The function names are kind of weird though.

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -286,35 +282,55 @@ static bool is_executable(const char *name)
+/// - if the file extension is in $PATHEXT and `name` is executable
+/// - if the result of any $PATHEXT extension appended to `name` is executable
+static bool is_executable_ext(char *name, char_u **abspath)
+  FUNC_ATTR_NONNULL_ARG(1, 2)
```
FUNC_ATTR_NONNULL_ARG(1)

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -286,35 +282,55 @@ static bool is_executable(const char *name)
 
     const char *ext_end = xstrchrnul(ext, ENV_SEPCHAR);
-    STRLCPY(buf_end, ext, ext_end - ext + 1);
+    size_t ext_len = (size_t)(ext_end - ext);
```
Probably we should handle `ext_len == 0` and `nameext_len == 0`.

erw7 committed:
fs.c: fix is_executable_ext()

- Corresponds to the case where pathext contains a zero-length
  extension.
- Remove unnecessary break statements.
- Fix function attributes.

justinmk committed:
test: "$PATHEXT=::"

justinmk commented:
Added a test case for `$PATHEXT=::` and other weird values. Will merge after CI.
